File: recruitment-server/recruitment-biz/src/main/java/com/andy/recruitment/biz/user/service/UserService.java
package com.andy.recruitment.biz.user.service;

import com.andy.recruitment.api.user.response.UserInfoRes;
import com.andy.recruitment.dao.user.entity.UserInfoDO;
import com.andy.spring.auth.LoginInfo;
import java.util.List;
import java.util.Map;

/**
 * 用户服务接口
 *
 * @author 庞先海 2020-02-02
 */
public interface UserService {

    /**
     * 注册用户信息
     *
     * @param userInfoDo 用户信息
     * @param operator   操作人
     * @return 用户ID
     */
    Long registerUser(UserInfoDO userInfoDo, String operator);

    /**
     * 通过手机号查询用户
     *
     * @param phone 手机号码
     * @return 用户信息
     */
    UserInfoRes getUserInfoByPhone(String phone);

    /**
     * 通过手机号查询用户
     *
     * @param phone 手机号码
     * @return 用户信息Do
     */
    UserInfoDO getUserInfoDoByPhone(String phone);

    /**
     * 根据用户id获取用户信息
     *
     * @param userId 用户ID
     * @return 用户信息
     */
    UserInfoRes getUserInfoByUserId(Long userId);

    /**
     * 根据用户id获取用户信息
     *
     * @param userId 用户ID
     * @return 用户信息
     */
    UserInfoDO getUserDoByUserId(Long userId);

    /**
     * 通过用户ID列表查询用户ID
     *
     * @param userIdList 用户已ID列表
     * @return 用户信息列表
     */
    List<UserInfoDO> getUserInfo(List<Long> userIdList);

    /**
     * 通过用户ID列表查询用户ID
     *
     * @param userIdList 用户已ID列表
     * @return 用户信息列表
     */
    Map<Long, UserInfoRes> getUserInfoRes(List<Long> userIdList);

    /**
     * 微信登录
     *
     * @param code 授权码
     * @return 用户信息
     */
    UserInfoDO wxLogin(String code);

    /**
     * 通过用户信息登录 所有信息都在userInfoDo 生成token和shirodeng登录
     *
     * @param userInfoDo 用户信息
     * @return 登录信息
     */
    LoginInfo userInfoLogin(UserInfoDO userInfoDo);

    /**
     * 更新用户信息
     *
     * @param userInfoDo 用户信息
     * @param operator   操作人
     */
    void updateUserInfo(UserInfoDO userInfoDo, String operator);

}

File: recruitment-go/biz/service/UserService.go
package service

import (
	"encoding/json"
	"errors"
	"fmt"
	"recruitment/biz/dao"
	"recruitment/common"
	"recruitment/constant"
	"recruitment/util"
	"sync"
)

type UserService struct {
	userDAO *dao.UserDAO
}

var userService *UserService
var userServiceLock sync.Mutex

func GetUserService() *UserService {
	if userService != nil {
		return userService
	}
	userServiceLock.Lock()
	defer userServiceLock.Unlock()
	if userService != nil {
		return userService
	}
	userService = &UserService{}
	userService.userDAO = dao.GetUserDAO()
	return userService
}

/*
登录信息 LoginRes
*/
type LoginInfo struct {
	UserId   uint   `json:"userId"`
	UserName string `json:"userName"`
	RealName string `json:"realName"`
	Phone    string `json:"phone"`
	//角色类型： 管理员 or 用户
	RoleType byte `json:"roleType"`
	//附加信息
	Attach string `json:"attach"`
	Status byte   `json:"status"`
}

type UserInfoRes struct {
	UserId   uint            `json:"userId"`
	Nickname string          `json:"nickname"`
	Phone    string          `json:"phone"`
	RealName string          `json:"realName"`
	Gender   common.BaseType `json:"gender"`
	Status   common.BaseType `json:"status"`
}

type UpdatePasswordReq struct {
	Phone          string `json:"phone"`
	SourcePassword string `json:"password" binding:"required"`
	NewPassword    string `json:"newPassword" binding:"required"`
}

func (this *UserService) GetCurrentUser(token string, userName string) (*LoginInfo, error) {
	decodeToken, err := util.Rc4DecodeByte(token, userName)
	if err != nil {
		return nil, err
	}
	var logInfo LoginInfo
	err = json.Unmarshal(decodeToken, &logInfo)
	if err != nil {
		return nil, err
	}
	if logInfo.UserName != userName {
		return nil, errors.New("unauthorized")
	}
	return &logInfo, nil
}

func (this *UserService) GetUserInfoByPhone(phone string) *UserInfoRes {
	userInfoDO := this.userDAO.GetUserInfoByPhone(phone)

	userInfoRes := UserInfoRes{
		UserId:   userInfoDO.ID,
		Nickname: userInfoDO.Nickname,
		Phone:    userInfoDO.Phone,
		RealName: userInfoDO.RealName,
		Gender:   common.BaseType{Code: userInfoDO.Gender, Desc: constant.GenderDict[userInfoDO.Gender]},
		Status:   common.BaseType{Code: userInfoDO.Status, Desc: constant.StatusDict[userInfoDO.Gender]},
	}

	return &userInfoRes
}

func (this *UserService) UpdateUserPassword(req *UpdatePasswordReq, operator string) (uint, error) {
	userInfoDO := this.userDAO.GetUserInfoByPhone(req.Phone)
	if userInfoDO.ID == 0 {
		return common.USER_NOT_EXIST, nil
	}
	fmt.Println("req c参数", req)
	fmt.Println("userDO ", userInfoDO)
	encodePass := util.HmacSha256Hash(req.SourcePassword, req.Phone)
	if encodePass != userInfoDO.Password {
		return common.USER_SOURCE_PASSWORD_ERROR, nil
	}
	encodeNewPass := util.HmacSha256Hash(req.NewPassword, req.Phone)
	updateDO := dao.UserInfoDO{
		BaseDO: common.BaseDO{
			ID: userInfoDO.ID,
		},
		Password: encodeNewPass,
	}
	return this.userDAO.UpdateUser(&updateDO, operator)
}
